### ISSUE TITLE
Is scanToken() swallowing tokens incorrectly?

### ISSUE THREAD

**Issue by vsajip** [javacc21/javacc21#89]
This Python code fails to parse:
```python
class Foo:
    thing = ['foo']

'''
First
'''


class Bar:
    thing = ['bar']


'''
Second
'''  # This comment causes a problem

class Baz:
    thing = ['baz']
```
The error:
```
Error processing file: /home/vinay/tmp/tests2.py
org.parsers.python.ParseException: 
Encountered an error at (or somewhere around) /home/vinay/tmp/tests2.py:17:1
Was expecting one of the following:
NEWLINE
Found string "class" of type CLASS
	at org.parsers.python.PythonParser.handleUnexpectedTokenType(PythonParser.java:22215)
	at org.parsers.python.PythonParser.consumeToken(PythonParser.java:22203)
	at org.parsers.python.PythonParser.SimpleStatement(PythonParser.java:857)
	at PythonParser.Statement(Python.javacc:63)
	at org.parsers.python.PythonParser.Statement(PythonParser.java:758)
	at PythonParser.PythonFile(Python.javacc:25)
	at org.parsers.python.PythonParser.PythonFile(PythonParser.java:230)
Parse failed on: /home/vinay/tmp/tests2.py
```
I'm not sure that the problem is in the grammar - it might well be to do with how `scanToken()` works, token chaining, unparsed tokens and all that good (!) stuff. When debugging, I set it up to break on the call to `PythonParser.Statement()` when the "Second" string literal is `lastConsumedToken.next`, then stopped at the next `PythonLexer.nextToken(Token tok)` - it's called from within `scanToken()` which is called from `check$Strings()`. Now, it loops around
```java
while (result== null) {
    nextTokenType= null;
    Token next= token_source.getNextToken();
    previous= next;
    if (!next.isUnparsed()) {
        result= next;
    }
    else if (next instanceof InvalidToken) {
        result= next.getNextToken();
    }
}
```
and appears to skip past the `COMMENT` and `NEWLINE` (because they are unparsed). The upshot is that the parser thinks that the `class` (for `Baz`) is the next token after the string literal, which of course fails because there should be a `NEWLINE` following the string literal. See this screenshot:
![ss3](https://user-images.githubusercontent.com/130553/126866784-fd5d4760-6d92-4feb-bfeb-762b89f0299e.png)
If that comment is removed, the file is parsed successfully.

**Comment by revusky** [comment]
Well, I'll try to figure this out. I'm not terribly happy with how this works, I have to admit. I recognize that it's overly complicated. It seems like there is a need for a refactoring such that the token lookahead caching/vending machinery does not rely on this token chaining.

**Comment by revusky** [comment]
It took me an inordinate amount of time to figure this out. The bug was in token chaining logic. Basically a newline is syntactically meaningful if it is on a line with something else that is syntactically meaningful, otherwise it is ignored by the parsing machinery, like whitespace in just about every other language. So, if you have:

        '''
           something
        ''' # comment [NewLine]

The newline is syntactically meaningful. So, the logic was that it got the previousToken via the chaining logic, and that's the comment, but then the previous token of the comment was (due to a tricky bug) not returning the previous string literal, but the DEDENT before it. So the (spurious) logic was that the newline was alone on the line with the comment and thus, not syntactically meaningful. So it was going on the the next token, CLASS, but the parsing machinery was expecting a NewLine.

I don't even like the fix I put in, but it will do for now.

**Comment by vsajip** [comment]
> It took me an inordinate amount of time to figure this out.

You found it more quickly than I would have :smile:

> Basically a newline is syntactically meaningful if it is on a line with something else that is syntactically meaningful

Does that mean multiple `NEWLINE`s are collapsed into one? I wondered about that. Why not just send it through as a `NEWLINE` even if there's nothing else (or just whitespace) on the line?

**Comment by revusky** [comment]
Well, as I said, my interpretation of things, my conceptual model if you will, has been that a NEWLINE is syntactically meaningful if it is on a line with other things that are syntactically meaningful. Thus, the newlines that terminate the following comment lines:  

   # comment1
   # comment2
   # comment3

are set to unparsed. Granted, the other possible interpretation is that the newline is part of the comment. In fact, as I recall, that was how I defined a comment initially, but then I had problems with that, so I decided that a comment was up to *but not including* the newline. 

Well, the official document on this is here: https://docs.python.org/3/reference/lexical_analysis.html and I just looked at that again, and they explain the whole concept quite differently. But I *think* it basically is equivalent to what I say above. They explain the thing in terms *logical* lines vs. *physical* lines. The NEWLINE at the end of a physical line that is *not* also a logical line is not syntactically meaningful, I guess. But what I'm doing is around here: https://github.com/javacc21/javacc21/blob/master/examples/python/PythonLexer.javacc#L86

The problem that I fixed before was due to the token chaining being broken, so the logic by which it decided whether the newline was meaningful or not was effectively broken. But I think it's okay now.

**Comment by vsajip** [comment]
Fixed in 6284c2a, thanks.
